autogpts/autogpt/autogpt/command_decorator.py

```diff
@@ -19,52 +13,58 @@
 ) -> Callable[[Callable[P, CO]], Callable[P, CO]]:
     """
-    The command decorator is used to create Command objects from ordinary functions.
+    The command decorator is used to add command metadata to ordinary functions.
+
+    Args:
+        names (list[str]): The names of the command.
+            If not provided, the function name will be used.
+        description (str): A brief description of what the command does.
+            If not provided, the first part of function docstring will be used.
+        parameters (dict[str, JSONSchema]): The parameters of the function
+            that the command executes.
+        is_valid (Callable[[CommandArgs], ValidityResult]):
+            A function that checks if the command with provided arguments is valid.
     """
 
     def decorator(func: Callable[P, CO]) -> Callable[P, CO]:
+        doc = func.__doc__ or ""
+        # If names is not provided, use the function name
+        command_names = names or [func.__name__]
+        # If description is not provided, use the first part of the docstring
+        command_description = description
+        if not command_description:
+            if not func.__doc__:
+                raise ValueError("Description is required if function has no docstring")
+            # Extract either before [Aa]rgs|[Rr]eturns: or everything
+            pattern = re.compile(r"^(.*?)(?:[Aa]rgs:|[Rr]eturns:)", re.DOTALL)
+            match = pattern.search(doc)
+            if match:
+                # Return the part of the docstring before "Args:" or "Returns:"
+                command_description = match.group(1).strip()
+            else:
+                # Return the entire docstring
+                command_description = doc.strip()
+
+        # Parameters
         typed_parameters = [
             CommandParameter(
                 name=param_name,
                 spec=spec,
             )
             for param_name, spec in parameters.items()
         ]
-        cmd = Command(
-            name=name,
-            description=description,
-            method=func,
-            parameters=typed_parameters,
-            enabled=enabled,
-            disabled_reason=disabled_reason,
-            aliases=aliases,
-            available=available,
-        )
-
-        if inspect.iscoroutinefunction(func):
-
-            @functools.wraps(func)
-            async def wrapper(*args: P.args, **kwargs: P.kwargs) -> Any:
-                return await func(*args, **kwargs)
-
-        else:
-
-            @functools.wraps(func)
-            def wrapper(*args: P.args, **kwargs: P.kwargs) -> Any:
-                return func(*args, **kwargs)
 
-        setattr(wrapper, "command", cmd)
-        setattr(wrapper, AUTO_GPT_COMMAND_IDENTIFIER, True)
+        setattr(func, "names", command_names)
+        setattr(func, "description", command_description)
+        setattr(func, "parameters", typed_parameters)
+        setattr(func, "is_valid", is_valid)
```

Comment:
This one was tricky because `func` was losing binding to `self` but I solve it with a custom `__get__` and it returns `Command` now + `from_decorated_function` is removed.